lib/fetch/index.js

```diff
@@ -740,34 +741,91 @@ function finalizeResponse (fetchParams, response) {
-  //    response consume given response and failure.on.
-  //    TODO
-  //    3. If response’s body is null, then queue a fetch task to run
-  //    processBody given null, with fetchParams’s task destination.on.
-  //    TODO
-  //    4. Otherwise, fully read response’s body given processBody,
-  //    processBodyError, and fetchParams’s task destination.on.
-  //    TODO
-
-  // TODO (spec): The spec doesn't specify this but we need to
-  // terminate fetch if we have an error response.
-  if (response.type === 'error') {
-    context.terminate({ reason: response.error })
+  // 4. If response’s body is null, then run processResponseEndOfBody.
+  if (response.body == null) {
+    processResponseEndOfBody()
+  }
+
+  // 5. Otherwise:
+  else {
+    TransformStream ??= require('stream/web').TransformStream
+    // 1. Let transformStream be a new a TransformStream.
+
+    // 2. Let identityTransformAlgorithm be an algorithm which, given chunk, enqueues chunk in transformStream.
+    const identityTransformAlgorithm = (chunk, controller) => {
+      controller.enqueue(chunk)
+    }
+
+    const transformStream = new TransformStream({
+      // 3. Set up transformStream with transformAlgorithm set to identityTransformAlgorithm and
+      //    flushAlgorithm set to processResponseEndOfBody.
+      transform: identityTransformAlgorithm,
+      flush: processResponseEndOfBody
+    })
+
+    // 4. Set response’s body to the result of piping response’s body through transformStream.
+    // Khafra note: although the spec is unclear, only readable streams can be piped through
+    if (response.body != null) {
+      response.body = response.body.pipeThrough(transformStream)
+    }
+  }
+
+  // 6. If fetchParams’s process response consume body is non-null, then:
+  if (fetchParams.processResponseConsumeBody != null) {
+    // 1. Let processBody given nullOrBytes be this step: run fetchParams’s process response consume body given response and nullOrBytes.
+    const processBody = (nullOrBytes) => {
+      fetchParams.processResponseConsumeBody(response, nullOrBytes)
+    }
+
+    // 2. Let processBodyError be this step: run fetchParams’s process response consume body given response and failure.
+    const processBodyError = (failure) => {
+      fetchParams.processResponseConsumeBody(response, failure)
+    }
+
+    // 3. If response’s body is null, then queue a fetch task to run processBody given null, with fetchParams’s task destination.
+    if (response.body == null) {
+      processBody(null)
+    }
+
+    // 4. Otherwise, fully read response’s body given processBody, processBodyError, and fetchParams’s task destination.
+    else {
+      // The discussion to use the arrayBuffer method over the spec's fully read body implementation
+      // can be found here https://github.com/nodejs/undici/pull/1147#discussion_r775847886
+      response.body.arrayBuffer().then(
+        processBody,
+        processBodyError
+      )
+    }
```

Comment:
We need to terminate fetch here or we will leak the connection:

```
else {
  // TODO (spec): The spec doesn't specify this but we need to
  // terminate fetch if we have an error response.
  if (response.type === 'error') {
    context.terminate({ reason: response.error })
}
```

Reply:
This should be resolved now 👍

Reply:
How?

Reply:
a4c9a1a8277041a8a6c9018b07c5fba61ecb658c

Reply:
Ah. That's fine.